File: backend/app/assistant.py
from __future__ import annotations

import re
import json
from dataclasses import dataclass
from typing import Literal

import httpx

from .config import OPENAI_API_KEY, OPENAI_MODEL
from .schemas import FilterRequest, MarkerFilters

Intent = Literal["ranking", "trend", "geography", "project_search"]
ChartType = Literal["bar", "line", "map", "table"]
GroupBy = Literal[
    "year",
    "organization_name",
    "donor_country",
    "country",
    "region",
    "region_macro",
    "sector_description",
    "subsector_description",
    "type_of_flow",
]


KNOWLEDGE_BASE = [
    {
        "title": "Disbursements and commitments",
        "body": "Disbursements are funds actually paid out. Commitments are pledged or approved funds that may not have been paid yet. Disbursements are the default metric because commitments are missing in many source rows.",
    },
    {
        "title": "Project-safe funding",
        "body": "Headline totals use v_project_amounts so repeated sector rows do not inflate project-level funding. Sector breakdowns use source sector rows to preserve multi-sector project detail.",
    },
    {
        "title": "Recipient geography",
        "body": "Recipient country, region, and macro region fields describe where funding is directed. Regional, global, bilateral, and unspecified recipients cannot always be placed on a country map.",
    },
    {
        "title": "Priority markers",
        "body": "Climate combines climate mitigation and climate adaptation markers. Environment combines environment, biodiversity, and desertification markers. Gender and nutrition use their own marker columns.",
    },
    {
        "title": "Search fields",
        "body": "Project search checks organization name, project title, project description, and reported channel name. It is useful for topics such as maternal health, infectious disease, or education.",
    },
]

COUNTRY_ALIASES = {
    "uk": "United Kingdom",
    "u.k.": "United Kingdom",
    "britain": "United Kingdom",
    "united kingdom": "United Kingdom",
    "us": "United States",
    "u.s.": "United States",
    "usa": "United States",
    "u.s.a.": "United States",
    "america": "United States",
    "china": "China (People's Republic of)",
}

class OpenAIPlannerError(RuntimeError):
    pass


@dataclass
class AskPlan:
    intent: Intent
    chart_type: ChartType
    group_by: GroupBy
    grain: Literal["project", "sector"]
    filters: FilterRequest
    limit: int
    confidence: float
    interpretation: list[str]
    context: list[dict[str, str]]
    planner: Literal["openai"] = "openai"


def build_plan(question: str, metadata: dict[str, list[str]]) -> AskPlan:
    if not OPENAI_API_KEY:
        raise OpenAIPlannerError("OPENAI_API_KEY is required for Ask Data.")
    try:
        return build_openai_plan(question, metadata)
    except (httpx.HTTPError, KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
        raise OpenAIPlannerError(f"OpenAI planner failed: {exc}") from exc


def build_openai_plan(question: str, metadata: dict[str, list[str]]) -> AskPlan:
    context = retrieve_context(question, limit=3)
    prompt = {
        "question": question,
        "retrieved_context": context,
        "allowed_values": _metadata_prompt(metadata),
    }
    payload = {
        "model": OPENAI_MODEL,
        "input": [
            {
                "role": "developer",
                "content": [
                    {
                        "type": "input_text",
                        "text": (
                            "You convert philanthropy dashboard questions into a strict query plan. "
                            "Use only allowed fields and values. Do not invent data. "
                            "Prefer disbursements unless the user asks for commitments. "
                            "Use project grain for headline, donor, recipient, geography, and trend questions. "
                            "Use sector grain only when grouping by sector or subsector. "
                            "Use searchText only for project-search/topic-search questions, not for marker words alone. "
                            "If the question is ambiguous, choose the safest broad plan with no invented filters."
                        ),
                    }
                ],
            },
            {"role": "user", "content": [{"type": "input_text", "text": json.dumps(prompt)}]},
        ],
        "text": {
            "format": {
                "type": "json_schema",
                "name": "dashboard_query_plan",
                "strict": True,
                "schema": {
                    "type": "object",
                    "additionalProperties": False,
                    "properties": {
                        "intent": {"type": "string", "enum": ["ranking", "trend", "geography", "project_search"]},
                        "chartType": {"type": "string", "enum": ["bar", "line", "map", "table"]},
                        "groupBy": {
                            "type": "string",
                            "enum": [
                                "year",
                                "organization_name",
                                "donor_country",
                                "country",
                                "region",
                                "region_macro",
                                "sector_description",
                                "subsector_description",
                                "type_of_flow",
                            ],
                        },
                        "grain": {"type": "string", "enum": ["project", "sector"]},
                        "metric": {"type": "string", "enum": ["disbursements", "commitments"]},
                        "years": {"type": "array", "items": {"type": "string"}},
                        "donorCountries": {"type": "array", "items": {"type": "string"}},
                        "recipientCountries": {"type": "array", "items": {"type": "string"}},
                        "organizations": {"type": "array", "items": {"type": "string"}},
                        "sectors": {"type": "array", "items": {"type": "string"}},
                        "flowTypes": {"type": "array", "items": {"type": "string"}},
                        "markers": {
                            "type": "object",
                            "additionalProperties": False,
                            "properties": {
                                "climate": {"type": ["boolean", "null"]},
                                "gender": {"type": ["boolean", "null"]},
                                "environment": {"type": ["boolean", "null"]},
                                "nutrition": {"type": ["boolean", "null"]},
                            },
                            "required": ["climate", "gender", "environment", "nutrition"],
                        },
                        "searchText": {"type": ["string", "null"]},
                        "confidence": {"type": "number"},
                    },
                    "required": [
                        "intent",
                        "chartType",
                        "groupBy",
                        "grain",
                        "metric",
                        "years",
                        "donorCountries",
                        "recipientCountries",
                        "organizations",
                        "sectors",
                        "flowTypes",
                        "markers",
                        "searchText",
                        "confidence",
                    ],
                },
            }
        },
        "max_output_tokens": 900,
    }
    with httpx.Client(timeout=12) as client:
        response = client.post(
            "https://api.openai.com/v1/responses",
            headers={"Authorization": f"Bearer {OPENAI_API_KEY}", "Content-Type": "application/json"},
            json=payload,
        )
        response.raise_for_status()
    raw_plan = json.loads(_response_text(response.json()))
    return _validated_openai_plan(raw_plan, metadata, context)


def _validated_openai_plan(raw: dict, metadata: dict[str, list[str]], context: list[dict[str, str]]) -> AskPlan:
    intent = _expect(raw.get("intent"), ("ranking", "trend", "geography", "project_search"), "intent")
    chart_type = _expect(raw.get("chartType"), ("bar", "line", "map", "table"), "chartType")
    group_by = _expect(raw.get("groupBy"), GroupBy.__args__, "groupBy")
    grain = _expect(raw.get("grain"), ("project", "sector"), "grain")
    metric = _expect(raw.get("metric"), ("disbursements", "commitments"), "metric")

    filters = FilterRequest(
        years=_filter_allowed(raw.get("years", []), metadata.get("years", [])),
        donorCountries=_resolve_values(raw.get("donorCountries", []), metadata.get("donorCountries", []), COUNTRY_ALIASES),
        recipientCountries=_resolve_values(raw.get("recipientCountries", []), metadata.get("recipientCountries", []), COUNTRY_ALIASES),
        organizations=_resolve_values(raw.get("organizations", []), metadata.get("organizations", []), {}),
        sectors=_resolve_values(raw.get("sectors", []), metadata.get("sectors", []), {}),
        flowTypes=_resolve_values(raw.get("flowTypes", []), metadata.get("flowTypes", []), {}),
        markers=MarkerFilters(**{key: raw.get("markers", {}).get(key) for key in ("climate", "gender", "environment", "nutrition")}),
        metric=metric,
        searchText=(raw.get("searchText") or None),
    )
    if intent != "project_search":
        filters.searchText = None
    interpretation = _interpretation(intent, chart_type, group_by, grain, filters)
    confidence = max(0.0, min(float(raw.get("confidence", 0.75)), 0.95))
    return AskPlan(intent, chart_type, group_by, grain, filters, 10, confidence, interpretation, context, planner="openai")


def retrieve_context(question: str, limit: int = 2) -> list[dict[str, str]]:
    tokens = set(_normalize(question).split())
    scored = []
    for doc in KNOWLEDGE_BASE:
        doc_tokens = set(_normalize(f"{doc['title']} {doc['body']}").split())
        scored.append((len(tokens & doc_tokens), doc))
    return [doc for score, doc in sorted(scored, key=lambda item: item[0], reverse=True)[:limit] if score > 0]


def _metadata_prompt(metadata: dict[str, list[str]]) -> dict[str, list[str]]:
    return {
        "years": metadata.get("years", []),
        "donorCountries": metadata.get("donorCountries", [])[:150],
        "recipientCountries": metadata.get("recipientCountries", [])[:220],
        "organizations": metadata.get("organizations", [])[:120],
        "sectors": metadata.get("sectors", [])[:120],
        "flowTypes": metadata.get("flowTypes", []),
        "markers": ["climate", "gender", "environment", "nutrition"],
        "metrics": ["disbursements", "commitments"],
        "groupBy": list(GroupBy.__args__),
    }


def _expect(value: object, allowed: tuple[str, ...], field: str) -> str:
    if isinstance(value, str) and value in allowed:
        return value
    raise ValueError(f"OpenAI returned invalid {field}: {value}")


def _response_text(data: dict) -> str:
    if isinstance(data.get("output_text"), str):
        return data["output_text"]
    chunks: list[str] = []
    for item in data.get("output", []):
        for content in item.get("content", []):
            if content.get("type") in {"output_text", "text"} and isinstance(content.get("text"), str):
                chunks.append(content["text"])
    if not chunks:
        raise ValueError("OpenAI response did not include text output")
    return "".join(chunks)


def _filter_allowed(values: object, allowed: list[str]) -> list[str]:
    if not isinstance(values, list):
        return []
    allowed_set = set(allowed)
    return _dedupe([str(value) for value in values if str(value) in allowed_set])


def _resolve_values(values: object, allowed: list[str], aliases: dict[str, str]) -> list[str]:
    if not isinstance(values, list):
        return []
    allowed_by_normalized = {_normalize(value): value for value in allowed}
    resolved: list[str] = []
    for value in values:
        text = str(value).strip()
        canonical = aliases.get(_normalize(text), text)
        if canonical in allowed:
            resolved.append(canonical)
            continue
        normalized = _normalize(canonical)
        if normalized in allowed_by_normalized:
            resolved.append(allowed_by_normalized[normalized])
    return _dedupe(resolved)


def _interpretation(
    intent: Intent,
    chart_type: ChartType,
    group_by: GroupBy,
    grain: Literal["project", "sector"],
    filters: FilterRequest,
) -> list[str]:
    parts = [
        f"Intent: {intent.replace('_', ' ')}",
        f"Chart: {chart_type}",
        f"Metric: {filters.metric}",
        f"Grouped by: {group_by}",
        f"Grain: {grain}",
    ]
    active_filters = []
    for label, values in (
        ("years", filters.years),
        ("donor countries", filters.donorCountries),
        ("recipient countries", filters.recipientCountries),
        ("organizations", filters.organizations),
        ("sectors", filters.sectors),
        ("flow types", filters.flowTypes),
    ):
        if values:
            active_filters.append(f"{label} = {', '.join(values)}")
    markers = [name for name, value in filters.markers.model_dump().items() if value is True]
    if markers:
        active_filters.append(f"markers = {', '.join(markers)}")
    if filters.searchText:
        active_filters.append(f"search = {filters.searchText}")
    parts.append("Filters: " + ("; ".join(active_filters) if active_filters else "none"))
    return parts


def _normalize(value: str) -> str:
    return re.sub(r"\s+", " ", value.casefold().replace("_", " ")).strip()


def _dedupe(values: list[str]) -> list[str]:
    result = []
    seen = set()
    for value in values:
        if value not in seen:
            seen.add(value)
            result.append(value)
    return result

File: api/_lib/assistant.js
const { HttpError } = require("./http");

const KNOWLEDGE_BASE = [
  {
    title: "Disbursements and commitments",
    body: "Disbursements are funds actually paid out. Commitments are pledged or approved funds that may not have been paid yet. Disbursements are the default metric because commitments are missing in many source rows.",
  },
  {
    title: "Project-safe funding",
    body: "Headline totals use v_project_amounts so repeated sector rows do not inflate project-level funding. Sector breakdowns use source sector rows to preserve multi-sector project detail.",
  },
  {
    title: "Recipient geography",
    body: "Recipient country, region, and macro region fields describe where funding is directed. Regional, global, bilateral, and unspecified recipients cannot always be placed on a country map.",
  },
  {
    title: "Priority markers",
    body: "Climate combines climate mitigation and climate adaptation markers. Environment combines environment, biodiversity, and desertification markers. Gender and nutrition use their own marker columns.",
  },
  {
    title: "Search fields",
    body: "Project search checks organization name, project title, project description, and reported channel name. It is useful for topics such as maternal health, infectious disease, or education.",
  },
];

const GROUP_BY = [
  "year",
  "organization_name",
  "donor_country",
  "country",
  "region",
  "region_macro",
  "sector_description",
  "subsector_description",
  "type_of_flow",
];

const COUNTRY_ALIASES = {
  uk: "United Kingdom",
  "u.k.": "United Kingdom",
  britain: "United Kingdom",
  "united kingdom": "United Kingdom",
  us: "United States",
  "u.s.": "United States",
  usa: "United States",
  "u.s.a.": "United States",
  america: "United States",
  china: "China (People's Republic of)",
};

function normalize(value) {
  return String(value ?? "").toLowerCase().replaceAll("_", " ").replace(/\s+/g, " ").trim();
}

function retrieveContext(question, limit = 3) {
  const tokens = new Set(normalize(question).split(" ").filter(Boolean));
  return KNOWLEDGE_BASE.map((doc) => {
    const docTokens = new Set(normalize(`${doc.title} ${doc.body}`).split(" ").filter(Boolean));
    return { score: [...tokens].filter((token) => docTokens.has(token)).length, doc };
  })
    .sort((a, b) => b.score - a.score)
    .filter((item) => item.score > 0)
    .slice(0, limit)
    .map((item) => item.doc);
}

function metadataPrompt(metadata) {
  return {
    years: metadata.years ?? [],
    donorCountries: (metadata.donorCountries ?? []).slice(0, 150),
    recipientCountries: (metadata.recipientCountries ?? []).slice(0, 220),
    organizations: (metadata.organizations ?? []).slice(0, 120),
    sectors: (metadata.sectors ?? []).slice(0, 120),
    flowTypes: metadata.flowTypes ?? [],
    markers: ["climate", "gender", "environment", "nutrition"],
    metrics: ["disbursements", "commitments"],
    groupBy: GROUP_BY,
    aliases: COUNTRY_ALIASES,
  };
}

async function buildPlan(question, metadata) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) throw new HttpError(503, "OPENAI_API_KEY is required for Ask Data.");

  const context = retrieveContext(question);
  const payload = {
    model: process.env.OPENAI_MODEL || "gpt-5-mini",
    input: [
      {
        role: "developer",
        content: [
          {
            type: "input_text",
            text:
              "You convert philanthropy dashboard questions into a strict query plan. Use only allowed fields and values. Do not invent data. Prefer disbursements unless the user asks for commitments. Use project grain for headline, donor, recipient, geography, and trend questions. Use sector grain only when grouping by sector or subsector. Use searchText only for project-search/topic-search questions, not for marker words alone. If the question is ambiguous, choose the safest broad plan with no invented filters.",
          },
        ],
      },
      {
        role: "user",
        content: [
          {
            type: "input_text",
            text: JSON.stringify({
              question,
              retrieved_context: context,
              allowed_values: metadataPrompt(metadata),
            }),
          },
        ],
      },
    ],
    text: {
      format: {
        type: "json_schema",
        name: "dashboard_query_plan",
        strict: true,
        schema: {
          type: "object",
          additionalProperties: false,
          properties: {
            intent: { type: "string", enum: ["ranking", "trend", "geography", "project_search"] },
            chartType: { type: "string", enum: ["bar", "line", "map", "table"] },
            groupBy: { type: "string", enum: GROUP_BY },
            grain: { type: "string", enum: ["project", "sector"] },
            metric: { type: "string", enum: ["disbursements", "commitments"] },
            years: { type: "array", items: { type: "string" } },
            donorCountries: { type: "array", items: { type: "string" } },
            recipientCountries: { type: "array", items: { type: "string" } },
            organizations: { type: "array", items: { type: "string" } },
            sectors: { type: "array", items: { type: "string" } },
            flowTypes: { type: "array", items: { type: "string" } },
            markers: {
              type: "object",
              additionalProperties: false,
              properties: {
                climate: { type: ["boolean", "null"] },
                gender: { type: ["boolean", "null"] },
                environment: { type: ["boolean", "null"] },
                nutrition: { type: ["boolean", "null"] },
              },
              required: ["climate", "gender", "environment", "nutrition"],
            },
            searchText: { type: ["string", "null"] },
            confidence: { type: "number" },
          },
          required: [
            "intent",
            "chartType",
            "groupBy",
            "grain",
            "metric",
            "years",
            "donorCountries",
            "recipientCountries",
            "organizations",
            "sectors",
            "flowTypes",
            "markers",
            "searchText",
            "confidence",
          ],
        },
      },
    },
    max_output_tokens: 900,
  };

  const response = await fetch("https://api.openai.com/v1/responses", {
    method: "POST",
    headers: { authorization: `Bearer ${apiKey}`, "content-type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new HttpError(503, `OpenAI planner failed: ${await response.text()}`);
  const json = await response.json();
  return validatePlan(JSON.parse(responseText(json)), metadata, context);
}

function responseText(json) {
  if (typeof json.output_text === "string") return json.output_text;
  const chunks = [];
  for (const item of json.output ?? []) {
    for (const content of item.content ?? []) {
      if ((content.type === "output_text" || content.type === "text") && typeof content.text === "string") {
        chunks.push(content.text);
      }
    }
  }
  if (!chunks.length) throw new HttpError(503, "OpenAI response did not include text output.");
  return chunks.join("");
}

function allowedList(values, allowed) {
  const allowedSet = new Set(allowed ?? []);
  return Array.isArray(values) ? [...new Set(values.map(String).filter((value) => allowedSet.has(value)))] : [];
}

function resolveList(values, allowed) {
  if (!Array.isArray(values)) return [];
  const exact = new Set(allowed ?? []);
  const byNormalized = new Map((allowed ?? []).map((value) => [normalize(value), value]));
  const result = [];
  for (const value of values) {
    const text = String(value).trim();
    const canonical = COUNTRY_ALIASES[normalize(text)] || text;
    if (exact.has(canonical)) result.push(canonical);
    else if (byNormalized.has(normalize(canonical))) result.push(byNormalized.get(normalize(canonical)));
  }
  return [...new Set(result)];
}

function validatePlan(raw, metadata, context) {
  const intent = expectOne(raw.intent, ["ranking", "trend", "geography", "project_search"], "intent");
  const chartType = expectOne(raw.chartType, ["bar", "line", "map", "table"], "chartType");
  const groupBy = expectOne(raw.groupBy, GROUP_BY, "groupBy");
  const grain = expectOne(raw.grain, ["project", "sector"], "grain");
  const metric = expectOne(raw.metric, ["disbursements", "commitments"], "metric");
  const filters = {
    years: allowedList(raw.years, metadata.years),
    donorCountries: resolveList(raw.donorCountries, metadata.donorCountries),
    recipientCountries: resolveList(raw.recipientCountries, metadata.recipientCountries),
    regions: [],
    macroRegions: [],
    organizations: resolveList(raw.organizations, metadata.organizations),
    sectors: resolveList(raw.sectors, metadata.sectors),
    subsectors: [],
    flowTypes: resolveList(raw.flowTypes, metadata.flowTypes),
    markers: raw.markers ?? {},
    metric,
    searchText: intent === "project_search" ? raw.searchText || null : null,
  };
  return {
    intent,
    chartType,
    groupBy,
    grain,
    filters,
    limit: 10,
    confidence: Math.max(0, Math.min(Number(raw.confidence ?? 0.75), 0.95)),
    interpretation: interpretation(intent, chartType, groupBy, grain, filters),
    context,
    planner: "openai",
  };
}

function expectOne(value, allowed, field) {
  if (typeof value === "string" && allowed.includes(value)) return value;
  throw new HttpError(503, `OpenAI returned invalid ${field}.`);
}

function interpretation(intent, chartType, groupBy, grain, filters) {
  const parts = [
    `Intent: ${intent.replaceAll("_", " ")}`,
    `Chart: ${chartType}`,
    `Metric: ${filters.metric}`,
    `Grouped by: ${groupBy}`,
    `Grain: ${grain}`,
  ];
  const active = [];
  for (const [label, key] of [
    ["years", "years"],
    ["donor countries", "donorCountries"],
    ["recipient countries", "recipientCountries"],
    ["organizations", "organizations"],
    ["sectors", "sectors"],
    ["flow types", "flowTypes"],
  ]) {
    if (filters[key]?.length) active.push(`${label} = ${filters[key].join(", ")}`);
  }
  const markers = Object.entries(filters.markers ?? [])
    .filter(([, value]) => value === true)
    .map(([key]) => key);
  if (markers.length) active.push(`markers = ${markers.join(", ")}`);
  if (filters.searchText) active.push(`search = ${filters.searchText}`);
  parts.push(`Filters: ${active.length ? active.join("; ") : "none"}`);
  return parts;
}

function summarizeAnswer(intent, metric, data) {
  if (!data?.length) return "I could not find matching rows for that question. Try broadening the filters or using a more general topic.";
  if (intent === "project_search") {
    const first = data[0];
    return `I found ${data.length} matching projects. The largest visible match is ${first.grant_recipient_project_title || "the top matching project"} for ${first.country || "an unspecified recipient"}.`;
  }
  if (intent === "trend") {
    const peak = data.reduce((best, item) => (Number(item.amount ?? 0) > Number(best.amount ?? 0) ? item : best), data[0]);
    const latest = data[data.length - 1];
    return `The ${metric} trend peaks in ${peak.label} at $${Number(peak.amount ?? 0).toLocaleString(undefined, { maximumFractionDigits: 1 })}M. The latest shown value is $${Number(latest.amount ?? 0).toLocaleString(undefined, { maximumFractionDigits: 1 })}M in ${latest.label}.`;
  }
  if (intent === "geography") {
    const first = data[0];
    return `The largest recipient geography is ${first.country} with $${Number(first.amount ?? 0).toLocaleString(undefined, { maximumFractionDigits: 1 })}M in ${metric}.`;
  }
  const first = data[0];
  return `The top result is ${first.label || "the top result"} with $${Number(first.amount ?? 0).toLocaleString(undefined, { maximumFractionDigits: 1 })}M in ${metric}.`;
}

module.exports = { buildPlan, summarizeAnswer };
